pkg/block/fetcher.go

```diff
@@ -566,4 +581,57 @@ func (f *ConsistencyDelayMetaFilter) Filter(metas map[ulid.ULID]*metadata.Meta,
 			delete(metas, id)
 		}
 	}
+
+	return nil
+}
+
+// IgnoreDeletionMarkFilter is a filter that filters out the blocks that are marked for deletion after a given delay.
+// The delay duration is to make sure that the replacement block can be fetched before we filter out the old block.
+// Delay is not considered when computing DeletionMarkBlocks map.
+// Not go-routine safe.
+type IgnoreDeletionMarkFilter struct {
+	logger          log.Logger
+	delay           time.Duration
+	bkt             objstore.BucketReader
+	deletionMarkMap map[ulid.ULID]*metadata.DeletionMark
+}
+
+// NewIgnoreDeletionMarkFilter creates IgnoreDeletionMarkFilter.
+func NewIgnoreDeletionMarkFilter(logger log.Logger, bkt objstore.BucketReader, delay time.Duration) *IgnoreDeletionMarkFilter {
+	return &IgnoreDeletionMarkFilter{
+		logger: logger,
+		bkt:    bkt,
+		delay:  delay,
+	}
+}
+
+// DeletionMarkBlocks returns block ids that were marked for deletion.
+func (f *IgnoreDeletionMarkFilter) DeletionMarkBlocks() map[ulid.ULID]*metadata.DeletionMark {
+	return f.deletionMarkMap
+}
+
+// Filter filters out blocks that are marked for deletion after a given delay.
+// It also returns the blocks that can be deleted since they were uploaded delay duration before current time.
+func (f *IgnoreDeletionMarkFilter) Filter(ctx context.Context, metas map[ulid.ULID]*metadata.Meta, synced GaugeLabeled, _ bool) error {
+	f.deletionMarkMap = make(map[ulid.ULID]*metadata.DeletionMark)
+
+	for id := range metas {
+		deletionMark, err := metadata.ReadDeletionMark(ctx, f.bkt, f.logger, id.String())
+		if err == metadata.ErrorDeletionMarkNotFound {
+			continue
+		}
+		if _, ok := errors.Cause(err).(metadata.ErrorUnmarshalDeletionMark); ok {
```

Comment:
After me make sentintel from it, we can  do 
```suggestion
		if _, ok := errors.Cause(metadata.ErrorUnmarshalDeletionMark); ok {
```

Reply:
I have made this change. Thanks for your suggestion.